src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> next, T value, Node<T> prev) {
+            this.value = value;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> nodeToAdd = new Node<>(null, value, null);
+        if (first == null && last == null) { // when list hasn't first and last
+            first = nodeToAdd;
+            last = nodeToAdd;
+        } else if (first != null && first.next == null) { // when list hasn't last
+            first.next = nodeToAdd;
+            nodeToAdd.prev = first;
+            last = nodeToAdd;
+        } else { // when list has first and last
+            nodeToAdd.prev = last;
+            last.next = nodeToAdd;
+            last = nodeToAdd;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        outOfBoundsCheckToAdd(index);
+        if (index == size) { // to add node to end of the list
+            add(value);
+        } else if (index <= size - 1) { // to add node in list
+            Node<T> nodeToAdd = new Node<>(null, value, null);
+            Node<T> nodeNext = findNodeByIndex(index);
+            Node<T> nodePrev = nodeNext.prev;
+            nodeToAdd.prev = nodeNext.prev;
+            nodeToAdd.next = nodeNext;
+            nodeNext.prev = nodeToAdd;
+            if (first == nodeNext) { // check if node on 0 index
+                first = nodeToAdd;
+            } else {
+                nodePrev.next = nodeToAdd;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        outOfBoundsCheckToGetSetRemove(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        outOfBoundsCheckToGetSetRemove(index);
+        Node<T> nodeToSet = findNodeByIndex(index);
+        T oldNodeValue = nodeToSet.value;
+        nodeToSet.value = value;
+        return oldNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        outOfBoundsCheckToGetSetRemove(index);
```

Comment:
move this to the `findNodeByIndex` method
```suggestion
```

Reply:
> move this to the `findNodeByIndex` method

i have to checks: outOfBoundsCheckToGetSetRemove(index) and outOfBoundsCheckToAdd(index), i check it out of the findNodeByIndex method to avoid situation where i need to add another if else?  to chose from this to checks. Should i still do it?

Reply:
any questions should be asked on Mate platform